src/dropout.jl

```diff
@@ -0,0 +1,156 @@
+        rng = _rng_from_array(A)
+        pT = convert(real(eltype(dst)), p)
+        _dropout!(rng, dst, src, pT, dims)
+    else
+        # This fast path isn't free, but no concerns about types changing:
+        copyto!(dst, src)
+    end
+end
+
+# This is the easy case in that we can safely use the output array for random numbers.
+function _dropout!(rng::AbstractRNG, dst::AbstractArray, src::AbstractArray, p::Real, dims::Colon)
+    T = real(eltype(dst))
+    val = convert(T, 1/(1-p))
+    rand!(rng, dst)
+    ## This is what we want, but it hits a SIMD bug, solved by _fast_broadcast!
+    # dst .= (dst.>p) .* val .* src
+    _fast_broadcast!(dst, src) do q, x
+        ((real(q)>p) * val) * x
+    end
+    dst
+end
+
+# For other dims, we we do need to allocate something.
+function _dropout!(rng::AbstractRNG, dst::AbstractArray, src::AbstractArray, p::Real, dims)
+    T = real(eltype(dst))
+    tmp = similar(dst, T, ntuple(d -> d in dims ? size(src,d) : 1, ndims(src)))
+    rand!(rng, tmp)
+    val = convert(T, 1/(1-p))
+    ## One-pass strategy -- faster on GPU
+    dst .= ((tmp.>p) .* val) .* src
+    ## Two-pass strategy -- slightly faster on some CPUs?
+    # _fast_broadcast!(tmp) do q
+    #     (q>p) * val
+    # end
+    # dst .= tmp .* src
+end
+
+# The gradient needs to keep the random choices made, thus store at least a BitArray,
+# but the following way turns out to be faster & simpler:
+function ChainRulesCore.rrule(::typeof(dropout), rng::AbstractRNG, A::AbstractArray, p::Real; dims = :)
+    T = float(real(eltype(A)))
+    val = convert(T, 1/(1-p))
+    keep = if dims isa Colon
+        similar(A, T) 
+    else
+        similar(A, T, ntuple(d -> d in dims ? size(A,d) : 1, ndims(A)))
+    end
+    rand!(rng, keep)
+    Y = @. ((keep>p) * val) * A
+    function dropout_back(Δ)
+        dY = unthunk(Δ)
+        dA = @. ((keep>p) * val) * dY
+        (NoTangent(), NoTangent(), dA, NoTangent())
+    end
+    return Y, dropout_back
+end
+# Possibly TODO: another approach to the gradient would be to copy the RNG
+# and then re-generate the same mask, instead of storing it. This saves memory
+# and seems about as fast, but needs a method `copy(::CUDA.RNG)` and careful checking.
+# https://github.com/FluxML/NNlib.jl/pull/454#issuecomment-1369357402
+
+"""
+    _fast_broadcast!(f, x, y, z...)
+
+This does `x .= f.(x, y, z...)`, but works around
+an issue with broadcasting that prevents SIMD in such cases.
+Can be removed once https://github.com/JuliaLang/julia/issues/43153 is fixed.
+
+Not intended for general use. Does not check sizes!
+"""
+function _fast_broadcast!(f::F, x::Array, yz...) where {F<:Function}
+    bc = Broadcast.instantiate(Broadcast.broadcasted(f, x, yz...))
+    @simd ivdep for I in eachindex(bc)
+        @inbounds x[I] = bc[I]
+    end
+    return x
+end
+function _fast_broadcast!(f::F, x::AbstractArray, yz...) where {F<:Function}
+    # CUDA does not suffer from this bug
+    broadcast!(f, x, x, yz...)
+end
```

Comment:
Move this to utils.jl?

Reply:
Can do. But maybe I vote to move it once there's a second use?